File: ios/LibFFT.m
#import "LibFFT.h"

@interface LibFFT ()

@property (nonatomic) NSInteger fftNLog;
@property (nonatomic) NSInteger fftN;
@property (nonatomic) double miny;
@property (nonatomic) NSMutableArray<NSNumber *> *real;
@property (nonatomic) NSMutableArray<NSNumber *> *imag;
@property (nonatomic) NSMutableArray<NSNumber *> *sinTable;
@property (nonatomic) NSMutableArray<NSNumber *> *cosTable;
@property (nonatomic) NSMutableArray<NSNumber *> *bitReverse;

@end

@implementation LibFFT
- (instancetype)initWithBufferSize:(NSInteger)bufferSize {
    self = [super init];
    if (self) {
        _fftNLog = round(log2(bufferSize));
        _fftN = 1 << _fftNLog;
        _miny = (_fftN << 2) * sqrt(2.0);

        _real = [NSMutableArray arrayWithCapacity:_fftN];
        _imag = [NSMutableArray arrayWithCapacity:_fftN];
        _sinTable = [NSMutableArray arrayWithCapacity:_fftN / 2];
        _cosTable = [NSMutableArray arrayWithCapacity:_fftN / 2];
        _bitReverse = [NSMutableArray arrayWithCapacity:_fftN];

        for (NSInteger i = 0; i < _fftN; i++) {
            [_real addObject:@(0.0)];
            [_imag addObject:@(0.0)];
        }

        // Populate _bitReverse array
        for (NSInteger i = 0; i < _fftN; i++) {
            NSInteger j = 0;
            for (NSInteger bit = 0; bit < _fftNLog; bit++) {
                if (i & (1 << bit)) {
                    j |= 1 << ((_fftNLog - 1) - bit);
                }
            }
            [_bitReverse addObject:@(j)];
        }

        for (NSInteger i = 0; i < _fftN / 2; i++) {
            double theta = i * 2 * M_PI / _fftN;
            [_cosTable addObject:@(cos(theta))];
            [_sinTable addObject:@(sin(theta))];
        }
    }
    return self;
}

- (NSArray<NSNumber *> *)transform:(NSArray<NSNumber *> *)inBuffer {
    NSMutableArray<NSNumber *> *outBuffer = [NSMutableArray arrayWithCapacity:_fftN / 2];

    NSInteger j0 = 1;
    NSInteger idx = _fftNLog - 1;
    double cosv, sinv, tmpr, tmpi;

    for (NSInteger i = 0; i < _fftN; i++) {
      NSInteger index = [_bitReverse[i] integerValue];
      _real[i] = @(i < inBuffer.count && index < inBuffer.count ? [inBuffer[index] floatValue] : 0.0);
        _imag[i] = @(0.0);
    }

    for (NSInteger i = _fftNLog; i != 0; i--) {
        NSInteger j = 0;
        while (j != j0) {
            cosv = [_cosTable[j * (1 << idx)] doubleValue];
            sinv = [_sinTable[j * (1 << idx)] doubleValue];
            NSInteger k = j;
            while (k < _fftN) {
                NSInteger ir = k + j0;
                tmpr = cosv * [_real[ir] doubleValue] - sinv * [_imag[ir] doubleValue];
                tmpi = cosv * [_imag[ir] doubleValue] + sinv * [_real[ir] doubleValue];
                _real[ir] = @([_real[k] doubleValue] - tmpr);
                _imag[ir] = @([_imag[k] doubleValue] - tmpi);
                _real[k] = @([_real[k] doubleValue] + tmpr);
                _imag[k] = @([_imag[k] doubleValue] + tmpi);
                k += j0 << 1;
            }
            j++;
        }
        j0 <<= 1;
        idx--;
    }

    for (NSInteger i = 0; i < _fftN / 2; i++) {
        double tmpr = [_real[i + 1] doubleValue];
        double tmpi = [_imag[i + 1] doubleValue];
        float value = round(sqrt(tmpr * tmpr + tmpi * tmpi));
        [outBuffer addObject:@(value)];
    }

    return outBuffer;
}
@end

File: android/src/main/java/com/audioparser/LibFFT.kt
package com.audioparser

import kotlin.math.*

class LibFFT(bufferSize: Int) {
    private val fftNLog: Int
    private val fftN: Int
    private val miny: Double
    private val real: MutableList<Double>
    private val imag: MutableList<Double>
    private val sinTable: MutableList<Double>
    private val cosTable: MutableList<Double>
    private val bitReverse: MutableList<Int>

    init {
        fftNLog = round(log(bufferSize.toDouble(), 2.0)).toInt()
        fftN = 1 shl fftNLog
        miny = (fftN shl 2) * sqrt(2.0)

        real = MutableList(fftN) { 0.0 }
        imag = MutableList(fftN) { 0.0 }
        sinTable = MutableList(fftN / 2) { 0.0 }
        cosTable = MutableList(fftN / 2) { 0.0 }
        bitReverse = MutableList(fftN) { 0 }

        var i: Int
        var j: Int
        var k: Int
        var reve: Int
        for (i in 0 until fftN) {
            k = i
            reve = 0
            for (j in 0 until fftNLog) {
                reve = reve shl 1 or (k and 1)
                k = k ushr 1
            }
            bitReverse[i] = reve
        }

        val dt = 2 * PI / fftN
        for (i in fftN / 2 - 1 downTo 1) {
            val theta = i * dt
            cosTable[i] = cos(theta)
            sinTable[i] = sin(theta)
        }
    }

    fun transform(inBuffer: FloatArray): FloatArray {
        val paddedInput = if (inBuffer.size < fftN) {
            FloatArray(fftN) { index ->
                if (index < inBuffer.size) inBuffer[index] else 0f
            }
        } else {
            inBuffer
        }

        var j0 = 1
        var idx = fftNLog - 1
        var cosv: Double
        var sinv: Double
        var tmpr: Double
        var tmpi: Double
        for (i in 0 until fftN) {
            real[i] = paddedInput[bitReverse[i]].toDouble()
            imag[i] = 0.0
        }

        var i = fftNLog
        while (i != 0) {
            var j = 0
            while (j != j0) {
                cosv = cosTable[j shl idx]
                sinv = sinTable[j shl idx]
                var k = j
                while (k < fftN) {
                    val ir = k + j0
                    tmpr = cosv * real[ir] - sinv * imag[ir]
                    tmpi = cosv * imag[ir] + sinv * real[ir]
                    real[ir] = real[k] - tmpr
                    imag[ir] = imag[k] - tmpi
                    real[k] += tmpr
                    imag[k] += tmpi
                    k += j0 shl 1
                }
                j++
            }
            i--
            j0 = j0 shl 1
            idx--
        }

        val outBuffer = FloatArray(fftN / 2)
        sinv = miny
        cosv = -miny
        for (i in fftN / 2 downTo 1) {
            tmpr = real[i]
            tmpi = imag[i]
            outBuffer[i - 1] = if (tmpr > cosv && tmpr < sinv && tmpi > cosv && tmpi < sinv) 0f else round(tmpr * tmpr + tmpi * tmpi).toFloat()
        }
        return outBuffer
    }

    val bufferSize: Int
        get() = fftN
}
